lib/ansible/plugins/inventory/aws_ec2.py

```diff
@@ -0,0 +1,635 @@
+    'network-interface.association.public-ip': ('NetworkInterfaces', 'Association', 'PublicIp'),
+    'network-interface.association.ip-owner-id': ('NetworkInterfaces', 'Association', 'IpOwnerId'),
+    'network-interface.association.allocation-id': ('ElasticGpuAssociations', 'ElasticGpuId'),
+    'network-interface.association.association-id': ('ElasticGpuAssociations', 'ElasticGpuAssociationId'),
+    'network-interface.attachment.attachment-id': ('NetworkInterfaces', 'Attachment', 'AttachmentId'),
+    'network-interface.attachment.instance-id': ('InstanceId',),
+    'network-interface.attachment.device-index': ('NetworkInterfaces', 'Attachment', 'DeviceIndex'),
+    'network-interface.attachment.status': ('NetworkInterfaces', 'Attachment', 'Status'),
+    'network-interface.attachment.attach-time': ('NetworkInterfaces', 'Attachment', 'AttachTime'),
+    'network-interface.attachment.delete-on-termination': ('NetworkInterfaces', 'Attachment', 'DeleteOnTermination'),
+    'network-interface.availability-zone': ('Placement', 'AvailabilityZone'),
+    'network-interface.description': ('NetworkInterfaces', 'Description'),
+    'network-interface.group-id': ('NetworkInterfaces', 'Groups', 'GroupId'),
+    'network-interface.group-name': ('NetworkInterfaces', 'Groups', 'GroupName'),
+    'network-interface.ipv6-addresses.ipv6-address': ('NetworkInterfaces', 'Ipv6Addresses', 'Ipv6Address'),
+    'network-interface.mac-address': ('NetworkInterfaces', 'MacAddress'),
+    'network-interface.network-interface-id': ('NetworkInterfaces', 'NetworkInterfaceId'),
+    'network-interface.owner-id': ('NetworkInterfaces', 'OwnerId'),
+    'network-interface.private-dns-name': ('NetworkInterfaces', 'PrivateDnsName'),
+    # 'network-interface.requester-id': (),
+    'network-interface.requester-managed': ('NetworkInterfaces', 'Association', 'IpOwnerId'),
+    'network-interface.status': ('NetworkInterfaces', 'Status'),
+    'network-interface.source-dest-check': ('NetworkInterfaces', 'SourceDestCheck'),
+    'network-interface.subnet-id': ('NetworkInterfaces', 'SubnetId'),
+    'network-interface.vpc-id': ('NetworkInterfaces', 'VpcId'),
+    'placement-group-name': ('Placement', 'GroupName'),
+    'platform': ('Platform',),
+    'private-dns-name': ('PrivateDnsName',),
+    'private-ip-address': ('PrivateIpAddress',),
+    'product-code': ('ProductCodes', 'ProductCodeId'),
+    'product-code.type': ('ProductCodes', 'ProductCodeType'),
+    'ramdisk-id': ('RamdiskId',),
+    'reason': ('StateTransitionReason',),
+    'root-device-name': ('RootDeviceName',),
+    'root-device-type': ('RootDeviceType',),
+    'source-dest-check': ('SourceDestCheck',),
+    'spot-instance-request-id': ('SpotInstanceRequestId',),
+    'state-reason-code': ('StateReason', 'Code'),
+    'state-reason-message': ('StateReason', 'Message'),
+    'subnet-id': ('SubnetId',),
+    'tag': ('Tags',),
```

Comment:
If you have
```
group_by:
  - tag-key=Name
```
in the config it creates the group `aws_ec2_tag_key_Name` for all instances that have that tag (regardless of the tag's value)
```
group_by:
  - tag:Name=Prod
```
creates the group `aws_ec2_tag_Name_Prod` for instances that have the Name: Prod tag.

Do you think it even make sense to make groups for a tag's key only? I don't know what the use case would be. If there isn't one, would you expect `tag-key=Name` to create a group for each of the values found (e.g. it might create `aws_ec2_tag_Name_Prod`, `aws_ec2_tag_Name_Dev`)?

Reply:
Yep, that's how ec2.py works by default if you include tags, so I'd assume *someone* will want that behavior in the new script.

Reply:
I was thinking 'constructed' features could do all this and we could ship an example 'ec2.yml' with them in such a way they emulate current behaviour ... that allows users to change them at any time w/o touching code ... also eliminates a lot of code

Reply:
I've stripped the group_by logic in https://github.com/ansible/ansible/pull/34416. If you're in favor of that, I'll focus my attention on that and add a few more tests (but I'm going to respond to the rest of the comments here, so I'm addressing things in one place). Constructed is a little weird with tags, but easier to fix that there later rather than merging and then needing to remove some of this code. 

aws_ec2 config:
```
plugin: aws_ec2
boto_profile: shertel
hostnames:
  - dns-name
  - tag:Name=Test
filters:
  instance-type: t2.micro
regions:
  - us-east-1
```
constructed config:
```
plugin: constructed
strict: False
keyed_groups:
  - prefix: arch
    key: 'Architecture'
    value: 'x86_64'
  - prefix: tag
    key: Tags
    value:
      "Key": "Name"
      "Value": "Test"
```

`ansible-inventory -i aws_hosts/testinv.aws_ec2.yml -i aws_hosts/constructed.config --graph`
(cannot just specify -i aws_hosts as a directory because constructed may be called first)
```
@all:
  |--@arch_x86_64:
  |  |--ec2-34-227-95-245.compute-1.amazonaws.com
  |--@aws_ec2:
  |  |--ec2-34-227-95-245.compute-1.amazonaws.com
  |--@tag____Key____Name____Value____Test___:
  |  |--ec2-34-227-95-245.compute-1.amazonaws.com
  |--@ungrouped:
```